### ISSUE TITLE
@:generic assert

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#4692]
With the latest compiler + compiler cache, I'm sometimes getting the following:

`File "codegen.ml", line 358, characters 104-110: Assertion failed`

It seems to come from generic build, I'm not sure what could have change here that can affect compilation.

**Comment by ncannasse** [comment]
Ok, I could locate the code involved, it comes from https://github.com/ncannasse/heaps/blob/master/hxd/impl/Partition.hx

**Comment by Simn** [comment]
Is there any heaps test or something I can compile to reproduce this?

**Comment by ncannasse** [comment]
I'm trying to come up with one

**Comment by ncannasse** [comment]
```haxe
class E {
	public var x : Float;
	public var y : Float;
	public var partChunk : Int;
	public var partNext : E;	
	public var partQueryNext : E;	
}

class Main {

    static function main() {
		var v = new hxd.impl.Partition<E>(512,512,4);
		trace(v.count(0,0,0,function(e) return true));
    }

}
```
Compile once, then fails when recompiled if Main.hx is modified

**Comment by Simn** [comment]
Thanks, I'll check it out. Seems strange that a Hashtbl in `typer_globals` would be affected by the compilation server at all.

**Comment by ncannasse** [comment]
Well, it is possible that some dependency is not well registered in generics, so when reusing a generic we don't add enough things into our context:

Here's the compiler output with `-v`
```
Processing Arguments [--cwd,D:\,-lib,heaps,-x,Main]
Reusing  cached module StdTypes
Reusing  cached module String
Reusing  cached module Array
Reusing  cached module neko.NativeArray
Reusing  cached module neko.Lib
Reusing  cached module haxe.io.Bytes
Reusing  cached module haxe.io.BytesData
Reusing  cached module neko.NativeString
Reusing  cached module haxe.Int64
Reusing  cached module Std
Reusing  cached module Class
Reusing  cached module Enum
Reusing  cached module neko.Boot
Reusing  cached module IntIterator
Reusing  cached module Math
Reusing  cached module haxe.Int32
Reusing  cached module haxe.Int64Helper
Reusing  cached module StringTools
Reusing  cached module StringBuf
Reusing  cached module haxe.io.Error
Reusing  cached module haxe.io.FPHelper
Reusing  cached module neko.vm.Tls
Reusing  cached module haxe.io.BytesInput
Reusing  cached module haxe.io.Input
Reusing  cached module haxe.io.Eof
Reusing  cached module haxe.io.BytesBuffer
Reusing  cached module haxe.EnumTools
Reusing  cached module EnumValue
Reusing  cached module Type
Reusing  cached module Reflect
Reusing  cached module haxe.Constraints
File d:\main.hx has been modified
Skipping cached module Main
Parsed d:\main.hx
Reusing  cached module hxd.impl.Partition
Reusing  cached module haxe.ds.Vector
Reusing  cached module hxd.Math
Skipping cached module hxd.impl.Partition_E(Main)
> File "codegen.ml", line 358, characters 104-110: Assertion failed
Stats = 1 files, 2 classes, 1 methods, 0 macros
Time spent : 0.073s
```

**Comment by Simn** [comment]
But shouldn't the Hashtbls in `typer_globals` be available anywhere?

**Comment by ncannasse** [comment]
It's on this line for me:
`let m = (try Hashtbl.find ctx.g.modules (Hashtbl.find ctx.g.types_module c.cl_path) with Not_found -> assert false) in`

**Comment by ncannasse** [comment]
For some reason hxd.impl.Partition is not found in the types_modules, while it's been marked as reused

**Comment by Simn** [comment]
Yes, it's the first lookup: `Hashtbl.find ctx.g.types_module c.cl_path`. I just don't understand how that can be the case because I thought the typer globals were, well, globals.

**Comment by ncannasse** [comment]
they're globals for the current compilation only

**Comment by ncannasse** [comment]
only reason is that module is marked as MSub ?

**Comment by ncannasse** [comment]
or the module path has been changed

**Comment by ncannasse** [comment]
Ok, that's the type which is not found, so it might have been removed from module then?

**Comment by ncannasse** [comment]
Uhm, it seems that types_module is not populated when reloading from compiler cache

**Comment by Simn** [comment]
It simply doesn't have an entry in `g.types_module`. I don't understand where that entry would come from if we don't recompile the thing. The only place `g.types_module` is modified is when we type the module, which we don't do when reusing it.

**Comment by ncannasse** [comment]
Fixed it: when reloading a module from compiler cache, we need to both add it to modules list and register its types paths in type_modules, which was not done before; I guess we didn't use types_module in generic building before?

**Comment by Simn** [comment]
We did, that's why I don't understand what exactly broke this. I'll keep this issue open to investigate, gotta make sure I didn't break this with some seemingly unrelated change.

**Comment by Simn** [comment]
Alright, I see. Turns out `add_module` [used to do that](https://github.com/HaxeFoundation/haxe/blob/3.2.0/typeload.ml#L2685) anyway. I have factored that part out during my work on allowing to type types into existing modules for #4160. What I didn't consider is that `add_module` is also called by the compilation server, so the declaration part was missing now.

That means your change is good. We don't have to do the redefinition/casing checks for types we pull out of the compilation server anyway.
